tests/components/songpal/test_media_player.py

```diff
@@ -0,0 +1,267 @@
+    await _call(media_player.SERVICE_TURN_OFF)
+    await _call(media_player.SERVICE_TOGGLE)
+    assert mocked_device.set_power.call_count == 3
+    mocked_device.set_power.assert_has_calls([call(True), call(False), call(False)])
+
+    await _call(media_player.SERVICE_VOLUME_SET, volume_level=0.6)
+    await _call(media_player.SERVICE_VOLUME_UP)
+    await _call(media_player.SERVICE_VOLUME_DOWN)
+    assert mocked_device.volume1.set_volume.call_count == 3
+    mocked_device.volume1.set_volume.assert_has_calls(
+        [call(60), call("+1"), call("-1")]
+    )
+
+    await _call(media_player.SERVICE_VOLUME_MUTE, is_volume_muted=True)
+    mocked_device.volume1.set_mute.assert_called_once_with(True)
+
+    await _call(media_player.SERVICE_SELECT_SOURCE, source="none")
+    mocked_device.input1.activate.assert_not_called()
+    await _call(media_player.SERVICE_SELECT_SOURCE, source="title1")
+    mocked_device.input1.activate.assert_called_once()
+
+    await hass.services.async_call(
+        songpal.DOMAIN,
+        SET_SOUND_SETTING,
+        {"entity_id": ENTITY_ID, "name": "name", "value": "value"},
+        blocking=True,
+    )
+    mocked_device.set_sound_settings.assert_called_once_with("name", "value")
+    mocked_device.set_sound_settings.reset_mock()
+
+    mocked_device2 = _create_mocked_device()
+    sys_info = MagicMock()
+    sys_info.macAddr = "mac2"
+    sys_info.version = SW_VERSION
+    type(mocked_device2).get_system_info = AsyncMock(return_value=sys_info)
+    entry2 = MockConfigEntry(
+        domain=songpal.DOMAIN, data={CONF_NAME: "d2", CONF_ENDPOINT: ENDPOINT}
+    )
+    entry2.add_to_hass(hass)
+    with _patch_media_player_device(mocked_device2):
+        await hass.config_entries.async_setup(entry2.entry_id)
+        await hass.async_block_till_done()
+
+    await hass.services.async_call(
+        songpal.DOMAIN,
+        SET_SOUND_SETTING,
+        {"entity_id": "all", "name": "name", "value": "value"},
+        blocking=True,
+    )
+    mocked_device.set_sound_settings.assert_called_once_with("name", "value")
+    mocked_device2.set_sound_settings.assert_called_once_with("name", "value")
+
+
+async def test_websocket_events(hass):
+    """Test websocket events."""
+    mocked_device = _create_mocked_device()
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    entry.add_to_hass(hass)
+
+    with _patch_media_player_device(mocked_device):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+    mocked_device.listen_notifications.assert_called_once()
+    assert mocked_device.on_notification.call_count == 4
+
+    notification_callbacks = mocked_device.notification_callbacks
+
+    volume_change = MagicMock()
+    volume_change.mute = True
+    volume_change.volume = 20
+    await notification_callbacks[VolumeChange](volume_change)
+    attributes = _get_attributes(hass)
+    assert attributes["is_volume_muted"] is True
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)